### DISCUSSION TITLE
How can I create a Reddit-like comments structure?

### DISCUSSION THREAD

**Discussion by intergalactisch** [staudenmeir/laravel-adjacency-list#235] [General]
I'm working on a website where there is a simple Post model, and a Polymorphic Comment relationship:
```
public function comments()
    {
        return $this->morphMany(Comment::class, 'commentable');
    }
```

and
```
$post->comments()
    ->latest()
    ->get()
    ->toTree;
```

These comments have a `parent_id`, making this package instantly awesome. With this, I want to be able to both paginate the main ancestor-tree, and also limit the amount of children it fetches to max. 5. 

How can I make this possible?

**Comment by staudenmeir** [comment]
Hi @intergalactisch,

> paginate the main ancestor-tree

You mean the root-level comments, right? Try this:

```php
$comments = $post->comments()
    ->isRoot()
    ->latest()
    ->with(['descendants' => fn($query) => $query->limit(5)])
    ->paginate();

$comments->each(
    fn(Comment $comment) => $comment->setRelation('children', $comment->descendants->toTree())
);
```

If you are using Laravel <11, you need an additional package to support the eager loading limit:
https://github.com/staudenmeir/eloquent-eager-limit

**Reply by intergalactisch** [reply]
This is awesome, thanks! 
Any performant way to also attach the total (not-limited) amount of available descendants, next to the limited items?

**Reply by staudenmeir** [reply]
Use `withCount()`:

```php
$post->comments()->withCount('descendants')->...
```

**Reply by intergalactisch** [reply]
Hmm, when I your first example I bump into this issue: 

```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'comments.parent_id' in 'group statement'
```

I am on "laravel/framework": "^11.0.3". Any idea why this happens?

Edit: when I use just a simple `->with([ 'descendants'])` it does work tho. I'm on MariaDB.

**Reply by staudenmeir** [reply]
Unfortunately, MariaDB doesn't yet support this kind of `withCount()` query.

There are two alternatives and you'll have to compare their performance.

You can count the total descendants with separate queries:

```php
$comments = $post->comments()
    ->isRoot()
    ->latest()
    ->with(['descendants' => fn($query) => $query->limit(5)])
    ->paginate();

$comments->each(
    fn(Comment $comment) => $comment->setRelation('children', $comment->descendants->toTree())
);

$comments->each(
    fn(Comment $comment) => $comment->setAttribute('descendants_count', $comment->descendants()->count())
);
```

Or you load all descendants and then filter them afterwards:

```php
$comments = $post->comments()
    ->isRoot()
    ->latest()
    ->with('descendants')
    ->paginate();

$comments->each(
    fn(Comment $comment) => $comment->setRelation('children', $comment->descendants->slice(0, 5)->toTree())
);                                                                                 ^^^^^^^^^^^^^

foreach ($comments as $comment) {
    dump($comment->descendants->count());
}
```
